src/Uno.UI/UI/Xaml/Controls/ListViewBase/VirtualizingPanelLayout.iOSmacOS.cs

```diff
@@ -0,0 +1,1696 @@
+			public const int ListViewBaseLayout_PrepareLayoutStop = 2;
+		}
+
+		private NativeListViewBase Owner => CollectionView as NativeListViewBase;
+
+		private enum DirtyState
+		{
+			// The state of the layout hasn't changed
+			None,
+			// The state of the layout has been completely invalidated
+			NeedsRelayout,
+			// The state of the layout has been invalidated because of an INotifyCollectionChanged operation
+			CollectionChanged,
+			// The state of the layout has been invalidated because the scroll has changed and sticky header positions need updating
+			NeedsHeaderRelayout
+		}
+
+		private enum UnusedSpaceState
+		{
+			/// <summary>
+			/// Initial state
+			/// </summary>
+			Unset,
+			/// <summary>
+			/// The panel left some available extent unused on a previous measure pass, and hasn't attempted to use it.
+			/// </summary>
+			HasUnusedSpace,
+			/// <summary>
+			/// The panel has called a relayout to try to claim unused space that was available on a previous measure.
+			/// </summary>
+			ConsumeUnusedSpacePending
+		}
+
+		#region Members
+		/// <summary>
+		/// All cached item <see cref="_LayoutAttributes"/>s, grouped by collection group.
+		/// </summary>
+		private readonly Dictionary<int, LayoutInfoDictionary> _itemLayoutInfos = new Dictionary<int, LayoutInfoDictionary>();
+		/// <summary>
+		/// All cached <see cref="_LayoutAttributes"/> for supplementary elements (Header, Footer, group headers), grouped by element type.
+		/// </summary>
+		private readonly Dictionary<string, LayoutInfoDictionary> _supplementaryLayoutInfos = new Dictionary<string, LayoutInfoDictionary>();
+		/// <summary>
+		/// The last element in the list. This is set when the layout is created (and will point to the databound size and position of 
+		/// the element, once it has been materialized).
+		/// </summary>
+		private _LayoutAttributes _lastElement;
+
+		/// <summary>
+		/// Locations of group header frames if they were to appear inline with items (ie not 'sticky').
+		/// </summary>
+		private readonly Dictionary<int, CGRect> _inlineHeaderFrames = new Dictionary<int, CGRect>();
+		protected readonly Dictionary<int, nfloat> _sectionEnd = new Dictionary<int, nfloat>();
+
+		private Thickness _padding;
+		private Dictionary<CachedTuple<int, int>, NSIndexPath> _indexPaths = new Dictionary<CachedTuple<int, int>, NSIndexPath>(CachedTuple<int, int>.Comparer);
+		private Dictionary<CachedTuple<int, int>, _LayoutAttributes> _layoutAttributesForIndexPaths = new Dictionary<CachedTuple<int, int>, _LayoutAttributes>(CachedTuple<int, int>.Comparer);
+		private DirtyState _dirtyState;
+		/// <summary>
+		/// The most recently returned desired size.
+		/// </summary>
+		private CGSize _lastReportedSize;
+		/// <summary>
+		/// The most recent available size given by measure.
+		/// </summary>
+		private CGSize _lastAvailableSize;
+		/// <summary>
+		/// The available size when layout was most recently recreated.
+		/// </summary>
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Refactor this method to add validation of parameter 'indexPath' before using it.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=4931087)